### PR TITLE
lib/perf: Implement perf on top of shmfs

### PR DESCRIPTION
**Pull request by pussuw**
Instead of using a global perf list (that does not work in protected builds) allocate the perf objects into shared memory.

### CODE REVIEW

**Comment by jlaitine** [comment]
Very nice and clean solution! This maybe increases the memory consumption for the flat build a bit, but I guess that shouldn't be too significant.

I added a few questions still though

File: src/lib/perf/perf_counter.cpp [hunk 1]
```diff
@@ -91,51 +111,124 @@ struct perf_ctr_interval : public perf_ctr_header {
...
 
-	if (ctr != nullptr) {
+	return 0;
+}
+
+static size_t perf_shmsize(int fd)
+{
+	struct stat sb;
+
+	if (fstat(fd, &sb) < 0) {
```
[the remark this answers is not in the document]
Comment (pussuw) [reply src/lib/perf/perf_counter.cpp:137]: It works now
https://github.com/tiiuae/nuttx/pull/213

File: src/lib/perf/perf_counter.cpp [hunk 2]
```diff
@@ -37,24 +37,44 @@
...
+#include <sys/mman.h>
+#include <sys/stat.h>
+
+#include <px4_platform_common/mmap.h>
 
 #include "perf_counter.h"
 
+#define PERF_SHMNAME_PREFIX	"_perf_"
+#define PERF_SHMNAME_STR	PERF_SHMNAME_PREFIX "%s"
+#define PERF_SHMNAME_MAX	64
```
[the remark this answers is not in the document]
Comment (pussuw) [reply src/lib/perf/perf_counter.cpp:62]: Sure thing

**Review by jlaitine — APPROVED** [review]

**Comment by pussuw** [comment]
Nuttx updated, it contains the necessary patches for shmfs (stat and truncate work now).

**Pull request merged**